[lldb] Add interface to check if UserExpression::Parse() is cacheable

augusto2112 opened the pull request:
When setting conditional breakpoints, we currently assume that a call to UserExpression::Parse() can be cached and resued multiple times. This may not be true for every user expression. Add a new method so subclasses of UserExpression can customize if they are parseable or not.

jimingham reviewed (approved):
Presumably this will do something for Swift expressions?  I don't think there's any reason that the expression for a location in C/C++/ObjC code would not be cacheable, so the fact that this is a no-op in all of llvm.org lldb is expected.

DavidSpickett commented:
This got forgotten it seems.

Could you add a comment in the header to address Jim's question?

So folks upstream know whether to care about overriding this for whatever new thing they might be building (there are lots of mysterious methods over in llvm and I end up ignoring them and hoping they're not important).

augusto2112 commented:
@DavidSpickett thanks for reminding me. I'll add a comment describing the situation. 

Right now, for conditional breakpoints, we assume that the condition expression is cacheable, since it's being run over and over again in the exact same context. This works for C/C++, but not for Swift, because Swift generics aren't monomorphized (the same problem would happen for other languages where generics aren't monomorphized either, so a more general escape hatch seemed appropriate to me).

To illustrate what I mean, a function such as  the following (in C++ syntax but Swift semantics):

```
template <class T>
void use(T t) {}

// In main
use(5);
use(std::string())
```

Is lowered only once as a truly generic function (unlike C++ which will generate one version per instantiation), so it isn't safe to cache the expression parse, as the argument type may be different in every invocation.

augusto2112 committed:
[lldb] Add interface to check if UserExpression::Parse() is cacheable

When setting conditional breakpoints, we currently assume that a call to
UserExpression::Parse() can be cached and resued multiple times. This
may not be true for every user expression. Add a new method so
subclasses of UserExpression can customize if they are parseable or not.
